### ISSUE TITLE
Fix deadlock in syncer

### ISSUE THREAD

**Issue by ChrisSchinnerl** [SiaFoundation/coreutils#100]
A random deadlock that prevents our syncer from shutting down.            

https://github.com/SiaFoundation/renterd/actions/runs/11030847304/job/30636343643?pr=1574

[stacktrace.txt](https://github.com/user-attachments/files/17144377/stacktrace.txt)

_Originally posted by @peterjan in https://github.com/SiaFoundation/coreutils/issues/92#issuecomment-2376111458_

**Comment by lukechampine** [comment]
Here's what I have so far:

The Syncer uses a WaitGroup to ensure that all its goroutines exit before `Syncer.Close` returns. The deadlock happens because `wg.Wait` is not returning. This could be caused by calling `wg.Add` more than `wg.Done`, but that doesn't seem to be the case; all occurrences of `wg.Add` are immediately followed by a deferred `wg.Done`. This suggests that one of the goroutines simply isn't returning.

Looking at the stack trace, we can see that there is indeed one outstanding goroutine: `573687`, which accepts an incoming connection and calls `runPeer`. This goroutine appears to be blocked on the mux's `AcceptStream` method. That's strange, because when `mux.Close` is called, `AcceptStream` should wake up, observe `m.err != nil`, and exit. Another strange thing is that the stack trace does not say that this goroutine has been blocked for very long. Maybe this is just an inconsistency in the trace; alternatively, it could mean that the goroutine is being woken repeatedly, but isn't exiting. This in turn would imply that `m.err == nil`, which would be *very* strange.

I think the next step is to add some debug logic to the `mux` package, and use it for `renterd`'s CI until we trigger the bug again. I'll get started on that.

**Comment by lukechampine** [comment]
Actually, first things first: `renterd` is using `mux@v1.2.0`, which is pretty outdated. Let's update it to `v1.3.0` and see if the bug recurs.

**Comment by ChrisSchinnerl** [comment]
> Actually, first things first: `renterd` is using `mux@v1.2.0`, which is pretty outdated. Let's update it to `v1.3.0` and see if the bug recurs.

I believe that this particular test run already uses `v1.3.0` since we updated the dependency on the `dev` branch 6 days ago and the PR was created 4 days ago.

**Comment by lukechampine** [comment]
I think not lol
```
created by go.sia.tech/mux/v2.newMux in goroutine 573687
    /home/runner/go/pkg/mod/go.sia.tech/mux@v1.2.0/v2/mux.go:377 +0x426
```

**Comment by lukechampine** [comment]
ok, so it's reproduced on v.1.30. Good to know.

**Comment by peterjan** [comment]
Looks like it's happening once in a blue moon though, encountered another:

https://github.com/SiaFoundation/renterd/actions/runs/11612827367/job/32338279894?pr=1643
[stacktrace.txt](https://github.com/user-attachments/files/17601514/1_test.ubuntu-latest.1.22.txt)

**Comment by ChrisSchinnerl** [comment]
@lukechampine seems like it's actually related to using the syncer after closing it. Which is maybe not an unlikely scenario in a multithreaded environment. We should probably return a good error when a user is trying to do that.

You can use the following code to reproduce the deadlock

```go
func TestSyncerDeadlock(t *testing.T) {
	log := zaptest.NewLogger(t)

	n, genesis := testutil.Network()
	store1, tipState1, err := chain.NewDBStore(chain.NewMemDB(), n, genesis)
	if err != nil {
		t.Fatal(err)
	}
	cm1 := chain.NewManager(store1, tipState1)

	store2, tipState2, err := chain.NewDBStore(chain.NewMemDB(), n, genesis)
	if err != nil {
		t.Fatal(err)
	}
	cm2 := chain.NewManager(store2, tipState2)

	l1, err := net.Listen("tcp", ":0")
	if err != nil {
		t.Fatal(err)
	}
	defer l1.Close()

	l2, err := net.Listen("tcp", ":0")
	if err != nil {
		t.Fatal(err)
	}
	defer l2.Close()

	s1 := syncer.New(l1, cm1, testutil.NewEphemeralPeerStore(), gateway.Header{
		GenesisID:  genesis.ID(),
		UniqueID:   gateway.GenerateUniqueID(),
		NetAddress: l1.Addr().String(),
	}, syncer.WithLogger(log.Named("syncer1")))
	go s1.Run(context.Background())

	s2 := syncer.New(l2, cm2, testutil.NewEphemeralPeerStore(), gateway.Header{
		GenesisID:  genesis.ID(),
		UniqueID:   gateway.GenerateUniqueID(),
		NetAddress: l2.Addr().String(),
	}, syncer.WithLogger(log.Named("syncer2")), syncer.WithSyncInterval(10*time.Millisecond))
	defer s2.Close()
	go s2.Run(context.Background())

	// connect s1 to s2 and close the connection
	fmt.Println("connect")
	if peer, err := s1.Connect(context.Background(), s2.Addr()); err != nil {
		t.Fatal(err)
	} else if err := peer.Close(); err != nil {
		t.Fatal(err)
	}

	fmt.Println("close s1")
	// close s1
	if err := s1.Close(); err != nil {
		t.Fatal(err)
	}

	fmt.Println("connect again")

	// try connecting again
	if _, err := s1.Connect(context.Background(), s2.Addr()); err != nil {
		t.Fatal(err)
	}
	fmt.Println("connected")

	// close s1 - this shouldn't deadlock
	s1.Close()

	fmt.Println("not reached")
}
```
